### ISSUE TITLE
Extend the confounder analysis framework with cohort identity and sequencing depth as covariates

### ISSUE THREAD

**Issue by SuggarCoder** [maximz/malid#7]
While reproducing Mal-ID on the Synapse dataset for an independent benchmarking  project, I performed a series of preprocessing audits that uncovered stronger batch signal in the source data than is currently quantified in the existing confounder framework. I want to share these findings as a **constructive extension** of the work you've already done (demographics regressed-out flavors, kBET, leave-one-cohort-out CV) — none of these audits contradict your existing analyses, but they do suggest two additional covariates that may be worth formalizing into the metamodel flavor system.

I am not claiming Mal-ID is invalid. The repository's existing batch diagnostics (`confounder_model.py`, `kbet_batch_evaluation.py`, 
`paired_sample_batch_effects.py`, `leave_one_cohort_out_metamodel.*.py`) are more careful than most AIRR ML papers do. But the covariates currently controlled for are `["age", "sex", "ethnicity_condensed"]`, and I believe the data warrants two additions: **`study_name`** and **a sequencing depth indicator**.

## TL;DR Findings

| Predictor | Features | Multiclass AUROC (3-fold CV using your published fold IDs) |
|---|---|---|
| Random / shuffled labels | n/a | 0.49 ✓ (sanity check passes) |
| **Raw line count of source `.tsv.gz` files** | 2 features: `log(n_BCR_lines)`, `log(n_TCR_lines)` | **0.735** |
| Cohort one-hot encoding | 13 features: `study_name` indicator | **0.980** |
| Per-isotype scalar features (log_n, SHM median, SHM mutated fraction) | 12 features | 0.940 |
| Mal-ID Model 1 reproduction (V-J PCA + SHM scalars, your spec) | 66 features | 0.916 (matches paper's 0.92) |

The first two rows are the relevant ones for this issue:

- **0.735 AUROC from raw file line counts alone**, with no parsing, no filtering, no dedup. This is upstream of any preprocessing in this repository.
- **0.980 AUROC from `study_name` one-hot alone**, before any sequence content is considered.

## Background: data heterogeneity across cohorts

Auditing the 593-patient Synapse dataset revealed substantial protocol heterogeneity across the 13 study cohorts. Two examples:

**Sequencing depth varies ~70× across cohorts** (raw BCR `.tsv.gz` line counts):
Covid19-buffycoat:    median 5,905,065 lines
Covid19-Seattle:      median 1,747,214 lines
Covid19-Stanford:     median    82,558 lines
Diabetes biobank:     median    59,517 lines

**Isotype coverage (`c_call` non-NaN rate) varies from 53% to 95%**:
HIV cohort:                 53%   (likely 5'RACE, no C-region coverage)
Covid19-buffycoat:          54%
Healthy-StanfordBC:         81%
Covid19-Stanford:           95%   (likely multiplex with C-region primers)

These differences appear to reflect protocol/primer choices made at each contributing site (5'RACE vs. multiplex, sample type, PCR cycles, depth 
allocation). They are not introduced by this repository's preprocessing — they exist in the raw `.tsv.gz` files on Synapse.

### Experiment 1: Pristine baseline (no preprocessing whatsoever)

```python
# Pseudo-code; full script available on request
import gzip, pandas as pd
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import roc_auc_score

# For each patient: count gzipped lines in their BCR + TCR raw .tsv.gz files
counts = []
for pid in metadata['participant_label']:
    bcr_n = sum(1 for _ in gzip.open(f"raw/BCR/part_table_{pid}.tsv.gz", 'rt')) - 1
    tcr_n = sum(1 for _ in gzip.open(f"raw/TCR/part_table_{pid}.tsv.gz", 'rt')) - 1
    counts.append({'pid': pid, 'bcr_raw_n': bcr_n, 'tcr_raw_n': tcr_n})

# Use ONLY log(bcr_n), log(tcr_n) as features
X = np.log1p(df[['bcr_raw_n', 'tcr_raw_n']].values)
y = df['disease']
folds = df['malid_cross_validation_fold_id_when_in_test_set']

# 3-fold CV using your published folds
# Result: AUROC 0.7350 ± 0.0102
```

**Result**: 0.735 AUROC. This is the lower bound of batch leakage in the source data — it is fully attributable to sequencing depth differences across cohorts, which themselves correlate with disease (since most cohorts contribute one disease class).

### Experiment 2: Cohort one-hot

13-dimensional one-hot encoding of `study_name`, no other features: **AUROC 0.9797 ± 0.0002**.

This near-deterministic prediction reflects the bijective nature of cohort × disease in this dataset (`HIV` cohort → mostly HIV patients, `Diabetes biobank` → mostly T1D, `Covid19-*` cohorts → all Covid19, etc.). The remaining 0.02 to a perfect 1.0 comes from the few mixed-disease cohorts (e.g., HIV cohort contains both HIV patients and African healthy controls).

### Experiment 3: Sanity check on the audit pipeline

Same pipeline with `--shuffle-labels`: AUROC 0.49–0.53 across all variants. The audit pipeline itself does not introduce inflation; the signal is in the 
data.

## Why this matters for the existing confounder framework

The current `train_metamodel.py` flavors include:
with_demographics_columns
demographics_regressed_out
demographics_only
demographics_only_with_interactions
isotype_counts_only

with `DEMOGRAPHICS_COLUMNS = ["age", "sex", "ethnicity_condensed"]`.

I want to suggest that two additional covariates may be worth adding to this framework:

1. **`study_name`** as a categorical covariate. Given the 0.98 AUROC achievable from cohort alone, controlling for cohort identity (in addition to 
   demographics) seems important for isolating immune-specific signal.

2. **A sequencing depth measure** (e.g., `log(num_clones_post_dedup)` or `log(num_raw_reads)`) as a continuous covariate. Given the ~70× variation in raw read counts across cohorts and the 0.735 AUROC achievable from depth alone, this covariate appears to carry substantial confound.

## Suggested action items (for discussion, not prescription)

If the maintainers find this useful, possible next steps:

1. **Add `cohort_regressed_out` flavor** to `get_metamodel_flavors()`, regressing out `study_name` (one-hot) from the metamodel feature matrix in 
   the same way `demographics_regressed_out` does.

2. **Add `depth_regressed_out` flavor**, regressing out `log(num_clones)` per locus.

3. **Add a "raw line count baseline" notebook** under `notebooks/confounder_*` that documents the 0.735 AUROC achievable from 
   pre-preprocessing file sizes alone. This is probably the cleanest articulation of the lower bound on batch signal in the source data.

4. **Update reporting language**: in the README and paper revisions, the sentence "demographics-only baseline achieves AUROC X" could be expanded to "demographics-only baseline X / cohort-only baseline 0.98 / depth-only baseline 0.735" so that readers can place the main model results against the full ladder of baselines.

## Why I'm raising this

I am building a Mal-ID-derived benchmark for protein language model methods (AbLang2, SCEPTR), and I want my evaluation framework to be appropriately cautious about claiming any AUROC improvement on top of the published 0.986. Without a `cohort_regressed_out` baseline, it's difficult to distinguish "PLM provides incremental immune signal" from "PLM also picks up cohort indicators". Adding such a flavor to the official Mal-ID toolkit would make benchmarks against Mal-ID more interpretable for the entire community.

If I've misread anything in the codebase, please let me know — I tried to trace through `train_repertoire_stats_model.py`, `repertoire_classifier.py`, 
`train_metamodel.py`, and the relevant notebooks before posting, but I may have missed something.

Thanks for releasing the data and the full pipeline — being able to do this kind of audit independently is only possible because the repo is unusually complete and well-documented for a Science paper.

---

**Software environment used for these experiments**: Python 3.10, scikit-learn 1.5, numpy 2.x, pandas 2.x, on the published Synapse dataset (`syn61987835`, accessed [DATE], all 593 BCR + 542 TCR `.tsv.gz` files).
